### PR TITLE
#336: Workflow optimistic locking

### PR DESCRIPTION
**Pull request by jozefbakus**
Implements #336

### CODE REVIEW

File: src/main/scala/za/co/absa/hyperdrive/trigger/persistance/WorkflowRepository.scala [hunk 1]
```diff
@@ -212,12 +215,11 @@ class WorkflowRepositoryImpl @Inject()(
   }
 
   override def updateWorkflow(workflow: WorkflowJoined, user: String)(implicit ec: ExecutionContext): Future[Unit] = {
-    val w = workflow.toWorkflow.copy(updated = Option(LocalDateTime.now()))
+    val version = workflow.version
+    val w = workflow.toWorkflow.copy(updated = Option(LocalDateTime.now()), version = workflow.version + 1)
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/persistance/WorkflowRepository.scala]: Don't you want to put the +1 logic into the `updateWithOptimisticLocking` method?

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/tableExtensions/optimisticLocking/OptimisticLockingTableQuery.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import scala.language.higherKinds
+
+trait OptimisticLockingTableQuery {
+  this: Profile with JdbcTypeMapper =>
+
+  import api._
+
+  implicit class OptimisticLockTableQueryExtension[T <: OptimisticLockingTable with AbstractTable[_]](tableQuery: TableQuery[T]) {
+
+    def updateWithOptimisticLocking(value: T#TableElementType, version: Long)(implicit ec: ExecutionContext): DBIOAction[Int, NoStream, Effect.Write] = {
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/tableExtensions/optimisticLocking/OptimisticLockingTableQuery.scala]: Is there a usage for this method?

**Review by kevinwallimann — COMMENTED** [review]
The error message in the Frontend is just "Sorry, something went wrong. Try again." That's not the right advice :-D

File: src/main/scala/za/co/absa/hyperdrive/trigger/persistance/WorkflowRepository.scala (hunk 1 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/persistance/WorkflowRepository.scala]: Yes, done, but it is not that nice

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/tableExtensions/optimisticLocking/OptimisticLockingTableQuery.scala (hunk 2 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/tableExtensions/optimisticLocking/OptimisticLockingTableQuery.scala]: Removed

**Comment by jozefbakus** [comment]
> The error message in the Frontend is just "Sorry, something went wrong. Try again." That's not the right advice :-D

Fixed

File: ui/src/app/constants/texts.constants.ts [hunk 3]
```diff
@@ -69,6 +69,7 @@ export const texts = {
   UPDATE_WORKFLOW_CONFIRMATION_CONTENT: 'Are you sure you want to update the workflow?',
   UPDATE_WORKFLOW_SUCCESS_NOTIFICATION: 'Workflow has been updated.',
   UPDATE_WORKFLOW_FAILURE_NOTIFICATION: 'Sorry, something went wrong. Try again.',
+  UPDATE_WORKFLOW_OPTIMISTIC_LOCKING_FAILURE_NOTIFICATION: 'Sorry, workflow was updated in the meantime. Please refresh and try again.',
```
Comment (kevinwallimann) [inline ui/src/app/constants/texts.constants.ts]: How about "Please refresh and apply your changes again"? Users might simply hit refresh and assume that their changes will somehow stay and thus lose all of their changes

Comment (jozefbakus) [reply ui/src/app/constants/texts.constants.ts]: Done

**Review by kevinwallimann — APPROVED** [review]

**Pull request merged**